src/payload.c

```diff
@@ -173,6 +178,9 @@ int payload_run(void)
     while (p)
         p = load_one_payload(p, 0);
 
+    if (macho_boot)
```

Comment:
this mixes an API for booting what I think is a raw macho payload with an API meant to boot a Linux kernel.

If we want to add support for this into m1n1 this needs more thoughts on where to put it and how to decide when we want to load a raw payload vs. a Linux kernel.

Reply:
> this mixes an API for booting what I think is a raw macho payload with an API meant to boot a Linux kernel.

I don't think payload.c is intended to be specific to Linux kernel payloads, but if it is, we can add another layer of indirection between `run_actions()` and `payload_run()`?

In my case, the macho payload is also a Linux kernel, plus the Corellium preloader.

> If we want to add support for this into m1n1 this needs more thoughts on where to put it and how to decide when we want to load a raw payload vs. a Linux kernel.

There's only ever one "real" payload, so I don't understand what would need to be decided.

Reply:
Sorry, my original comment was a little bit confusing.

Right now payload.c loads various payloads, prepares the FDT, spins up the other cores and jumps to a payload which has a linux-compatible header.

Now you want to introduce support for "raw" payloads that don't have a linux-kernel header and neither need the FDT nor the other cores nor support any other payloads so far.

I'm not sure we want to support this in the long term since we won't need the Corellium preloader anymore in the near future. But if we (read: marcan ;)) wants to support this we should think a little bit more where to put that code. I feel like it should be separate from the normal payload handling since it's doing something very different.

Reply:
> Sorry, my original comment was a little bit confusing.

No need to apologize. Again, thank you for your comments!

> Right now payload.c loads various payloads

Yes. That's what I want to do for macho payloads as well.

> prepares the FDT,

No, kboot.c does that.

> spins up the other cores

That should also be moved to kboot.c, I think.

> and jumps to a payload

It'd still do that.

> which has a linux-compatible header.

... but this requirement would be relaxed. In fact, it already is relaxed, since we support compressed payloads.

> Now you want to introduce support for "raw" payloads that don't have a linux-kernel header and neither need the FDT nor the other cores nor support any other payloads so far.

Yes. It's a different kind of payload, so it doesn't use kboot.c, but it's still a payload, so it uses payload.c. You're right that, right now, macho payloads cannot be combined with other payloads, so that part of the payload.c code isn't useful yet, but decompression should work...

> I'm not sure we want to support this in the long term since we won't need the Corellium preloader anymore in the near future. 

I think it would be impolite to go out of our way in order not to support booting the Corellium kernels.

(As for what we want in the long term, I'm afraid my ideas are a bit different. m1n1 is a good thing and possibly necessary to get us there, but I would much prefer to boot to a Linux kernel .macho image directly, load an initrd, and kexec() the real kernel from there, without using m1n1 in the ordinary boot process. In fact, I'd like to go the reverse route and prepare a kexec()-able m1n1 image that we can boot from Linux for experimentation and compatibility...)

> But if we (read: marcan ;)) wants to support this we should think a little bit more where to put that code. I feel like it should be separate from the normal payload handling since it's doing something very different.

I don't see why we need another layer of abstraction: there are payloads, handled by payload.c; macho images are one (handled by macho.c), Linux images are another (handled by kboot.c), and possibly booting MacOS or Windows would require a third type of payload (in the case of MacOS, we probably want to override the ADT (which would be another payload type) so we can keep some hardware for our nefarious purposes, but I'm not even sure we can boot to MacOS at all)).

Another type of payload which I think would be useful would be a simple if-then payload, so we can switch between payloads based, for example, on which USB port is connected or whether we're on AC power.

```
cat m1n1.macho if-no-usb-connected.payload linux.macho
```

Sorry this got a bit long: in summary, I think we should treat macho payloads as "normal" payloads, and move the Linux-Image-specific stuff to kboot.c. We should provide a way for people to use the Corellium kernel (which is very usable) or switch to m1n1 for development without going through the long-press-`bputil` process every time.

Reply:
> 
> > I'm not sure we want to support this in the long term since we won't need the Corellium preloader anymore in the near future.
> 
> I think it would be impolite to go out of our way in order not to support booting the Corellium kernels.

Not adding support for something because it makes the code more complex is hardly the same as going out of our way to not support booting these kernels.

> 
> Sorry this got a bit long: in summary, I think we should treat macho payloads as "normal" payloads, and move the Linux-Image-specific stuff to kboot.

Exactly! All I'm saying is that right now payload and kboot are rather strongly coupled and if we want to support different kinds of chain load sequences (e.g. with fdt/smp vs. just raw macho/elf/whatever file) we should actually separate this.

>We should provide a way for people to use the Corellium kernel (which is very usable) or switch to m1n1 for development without going through the long-press-`bputil` process every time.

Don't get me wrong, I completely agree that it's useable. It's just that we will make m1n1 more complex and potentially create more maintenance work in the future for something that might be obsolete soon. 
No one really knows what Corellium's plans are and for all we know they won't release another kernel.


I'd definitely wait for marcan to decide if he wants to support these kinds of payloads.

Reply:
> > > I'm not sure we want to support this in the long term since we won't need the Corellium preloader anymore in the near future.
> > I think it would be impolite to go out of our way in order not to support booting the Corellium kernels.
> Not adding support for something because it makes the code more complex is hardly the same as going out of our way to not support booting these kernels.

I'm sorry I worded that so strongly. This is hardly the right place to discuss project governance and the attitude toward Stan@Corellium's work.

> > Sorry this got a bit long: in summary, I think we should treat macho payloads as "normal" payloads, and move the Linux-Image-specific stuff to kboot.
> 
> Exactly! All I'm saying is that right now payload and kboot are rather strongly coupled and if we want to support different kinds of chain load sequences (e.g. with fdt/smp vs. just raw macho/elf/whatever file) we should actually separate this.

I'm still not sure what kind of separation you're proposing. Moving stuff from payload.c to kboot.c is hardly a big deal?

> > We should provide a way for people to use the Corellium kernel (which is very usable) or switch to m1n1 for development without going through the long-press-`bputil` process every time.
> Don't get me wrong, I completely agree that it's useable. It's just that we will make m1n1 more complex and potentially create more maintenance work in the future for something that might be obsolete soon.

I'm sorry if my code made it seem like macho is a complex format. It's not. The cost-benefit tradeoffs seem quite clear-cut to me.

> No one really knows what Corellium's plans are and for all we know they won't release another kernel.

"For all we know they won't" sounds like you have good reason to believe they won't. Is that the case? Anyway, even in that case, I think Stan's code is a much better starting point than starting over from scratch...

> I'd definitely wait for marcan to decide if he wants to support these kinds of payloads.

I'd definitely appreciate his comments or opinions.

Reply:
> > 
> > 
> > Exactly! All I'm saying is that right now payload and kboot are rather strongly coupled and if we want to support different kinds of chain load sequences (e.g. with fdt/smp vs. just raw macho/elf/whatever file) we should actually separate this.
> 
> I'm still not sure what kind of separation you're proposing. Moving stuff from payload.c to kboot.c is hardly a big deal?

No, I completely agree with you! Moving stuff from payload.c to kboot.c sounds like a good solution if we want to support these kind of payloads.

> 
> > > We should provide a way for people to use the Corellium kernel (which is very usable) or switch to m1n1 for development without going through the long-press-`bputil` process every time.
> > > Don't get me wrong, I completely agree that it's useable. It's just that we will make m1n1 more complex and potentially create more maintenance work in the future for something that might be obsolete soon.
> 
> I'm sorry if my code made it seem like macho is a complex format. It's not. The cost-benefit tradeoffs seem quite clear-cut to me.

The code implements a subset of macho. macho itself is much more complex which is what I'm afraid of.

> 
> > No one really knows what Corellium's plans are and for all we know they won't release another kernel.
> 
> "For all we know they won't" sounds like you have good reason to believe they won't. Is that the case? 

Wasn't meant to sound that way.

>Anyway, even in that case, I think Stan's code is a much better starting point than starting over from scratch...

Agreed.